### PR TITLE
Css modules theming

### PR DESCRIPTION
**Pull request by Tri4Ty**

### CODE REVIEW

File: src/Link.js [hunk 1]
```diff
@@ -21,18 +31,13 @@ const Link = props => {
 export default Link;
 
 Link.propTypes = {
-  id: function(props, propName, componentName) {
```
Comment (sali1982) [inline src/Link.js]: same question as LinkGroup

File: src/LinkGroup.js [hunk 2]
```diff
@@ -1,40 +1,43 @@
...
+import Link from './Link';
+import styles from './LinkGroup.css';
 
 class LinkGroup extends Component {
   constructor(props) {
     super(props);
     this.linkSelection = this.linkSelection.bind(this);
 
     this.state = {
-      selected: undefined
```
Comment (sali1982) [inline src/LinkGroup.js]: actually I mean here, do not we want to have the first one selected by default?

File: src/LinkGroup.js [hunk 3]
```diff
@@ -1,40 +1,43 @@
...
+      selected: undefined,
     };
   }
 
   componentDidMount() {
-    let { selected } = this.props;
+    const { selected } = this.props;
```
Comment (sali1982) [inline src/LinkGroup.js]: is there any need to change this to const? we mostly use let. If there is a good reason please share it.

File: src/LinkGroup.js [hunk 4]
```diff
@@ -44,20 +47,22 @@ class LinkGroup extends Component {
...
-        />
-      );
+    const { links } = this.props;
+    const { selected } = this.state;
+    const theme = composeThemeFromProps(styles, this.props, {
+      compose: 'Replace',
     });
-    return <ul className="react-link-group">{linkList}</ul>;
+
+    const linkList = links.map((link) => (
```
Comment (sali1982) [inline src/LinkGroup.js]: there is no need to return anymore?

File: src/LinkGroup.js [hunk 5]
```diff
@@ -66,20 +71,15 @@ export default LinkGroup;
 LinkGroup.propTypes = {
   links: PropTypes.arrayOf(
     PropTypes.shape({
-      id: function(props, propName, componentName) {
```
Comment (sali1982) [inline src/LinkGroup.js]: why the id had an function in the shape and now it is gone?

File: webpack.core.config.js [hunk 6]
```diff
@@ -28,7 +28,17 @@ module.exports = {
       },
       {
         test: /\.(css)$/,
-        use: ["style-loader", "css-loader"]
+        use: [
+          "style-loader",
```
Comment (sali1982) [inline webpack.core.config.js]: is this the new way to have css-loader? what is options for?

File: src/LinkGroup.js [hunk 7]
```diff
@@ -44,20 +47,22 @@ class LinkGroup extends Component {
...
-        <Link
-          key={link.id}
-          {...link}
-          selected={link.id === selected}
-          selectionCallback={this.linkSelection}
-        />
-      );
+    const { links } = this.props;
+    const { selected } = this.state;
+    const theme = composeThemeFromProps(styles, this.props, {
```
Comment (sali1982) [inline src/LinkGroup.js]: what does composeThemeFromProps do? and how would it know to have reactLinkGroup for theme! which we see a few lines below? does it add Theme to all the props?

File: dev/src/App.js [hunk 8]
```diff
@@ -1,29 +1,46 @@
...
 
     this.state = {
-      selected: undefined
+      selected: undefined,
+      themeSelected: undefined,
     };
   }
 
+  selectedTheme(id) {
+    this.setState({ themeSelected: id });
```
Comment (sali1982) [inline dev/src/App.js]: it is confusing for me. Is it really the themeSelected id? or this dropdown has some theme? This name makes me think at the beggining that is the state of the theme which is selected!

File: .prettierrc [hunk 9]
```diff
@@ -0,0 +1,5 @@
+{
+  "singleQuote": true,
```
Comment (sali1982) [inline .prettierrc]: I was under impresssion we do not want to have single, 
Can you please explain why we had to change all to single?!

File: .eslintrc [hunk 10]
```diff
@@ -5,7 +5,9 @@
     ],
     "rules": {
         "max-len": [1,120, 2, {"ignoreComments": true}],
-        "react/prop-types": [2]
+        "react/prop-types": [2],
+        "linebreak-style": ["error", "windows"],
```
Comment (sali1982) [inline .eslintrc]: why did you add this rule?

File: .travis.yml [hunk 11]
```diff
@@ -1,3 +1,3 @@
 language: node_js
 node_js:
-  - "8.12.0"
```
[the remark this answers is not in the document]
Comment (Tri4Ty) [reply .travis.yml]: these files are ignored by prettier and aslant

File: .storybook/webpack.config.js [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+    rules: [
+      {
+        test: /\.css$/,
+        loaders: [
+          require.resolve("style-loader"),
+          {
+            loader: require.resolve("css-loader"),
+            options: {
+              modules: {
+                localIdentName: "[name]__[local]___[hash:base64:5]"
```
[the remark this answers is not in the document]
Comment (Tri4Ty) [reply .storybook/webpack.config.js]: not managed by prettier or eslint

File: dev/src/config.js [hunk 13]
```diff
@@ -1,24 +1,24 @@
 const LINKS_CONFIG = [
   {
-    id: "_CLEAR_",
```
[the remark this answers is not in the document]
Comment (Tri4Ty) [reply dev/src/config.js]: it's a test file... we aren't using _clear_ in the application so I changed it to what we use (null)

File: package.json [hunk 14]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "react-link-group",
-  "version": "1.1.7",
```
[the remark this answers is not in the document]
Comment (Tri4Ty) [reply package.json]: changed to css modules... considering it major

File: package.json [hunk 15]
```diff
@@ -81,7 +81,10 @@
     "webpack-dev-server": "^3.7.2",
     "webpack-merge": "^4.2.1"
   },
-  "dependencies": {},
+  "dependencies": {
+    "@css-modules-theme/react": "^2.2.0",
```
[the remark this answers is not in the document]
Comment (Tri4Ty) [reply package.json]: you should do some reading about package.json and get a basic understanding of what dependencies and peer-dependencies are ... this is pretty basic and designers working with npm and package.json files should understand this ... it's fundamentally how npm works (https://docs.npmjs.com/files/package.json)

File: src/LinkGroup.js (hunk 7 above)
Comment (Tri4Ty) [reply src/LinkGroup.js]: here you go... you'll have to read the documentation to understand how the css-modules-theme works: https://github.com/klimashkin/css-modules-theme

File: webpack.core.config.js (hunk 6 above)
Comment (Tri4Ty) [reply webpack.core.config.js]: great questions... here's where you can learn about it: https://github.com/css-modules/css-modules ... simple configuration on how it's done

File: dev/src/App.js (hunk 8 above)
Comment (Tri4Ty) [reply dev/src/App.js]: test file, just making up names in the state to separate which tab is select for the themed version and non-themed version. this code is not used in the component so naming was arbitrary

File: .prettierrc (hunk 9 above)
Comment (Tri4Ty) [reply .prettierrc]: this is prettier... I'm telling prettier to accept the single quote cause the new version of eslint enforced it

File: .eslintrc (hunk 10 above)
Comment (Tri4Ty) [reply .eslintrc]: cause new version of eslint was enforcing new lines and there is no way to configure prettier to ignore it so I disabled the rule ... otherwise every time you build you will need to go fix this lint error

File: src/Link.js [hunk 16]
```diff
@@ -1,17 +1,27 @@
...
     extraData = ` (${value})`;
   }
 
+  const theme = composeThemeFromProps(styles, props, { compose: 'Replace' });
+
   return (
-    <li className={linkClass} onClick={() => selectionCallback(id)}>
+    <li
+      className={
+        selected ? cx(theme.reactLink, theme.selected) : theme.reactLink
```
[the remark this answers is not in the document]
Comment (Tri4Ty) [reply src/Link.js]: here's a link to Google... https://github.com/jamesknelson/react-cx

File: dev/src/App.js [hunk 17]
```diff
@@ -1,29 +1,46 @@
-import React, { Component } from "react";
-import LinkGroup from "../../src/LinkGroup";
-import LINKS_CONFIG from "./config";
-import "./styles.css";
+import React, { Component } from 'react';
+
+import LinkGroup from '../../src/LinkGroup';
+import LINKS_CONFIG from './config';
+
+import theme from './TestTheme.css';
```
[the remark this answers is not in the document]
Comment (Tri4Ty) [reply dev/src/App.js]: :) cause that's the theme we are testing this with

File: dev/src/App.js [hunk 18]
```diff
@@ -1,29 +1,46 @@
...
 
 class App extends Component {
   constructor(props) {
     super(props);
     this.selected = this.selected.bind(this);
+    this.selectedTheme = this.selectedTheme.bind(this);
 
     this.state = {
-      selected: undefined
+      selected: undefined,
```
[the remark this answers is not in the document]
Comment (Tri4Ty) [reply dev/src/App.js]: we could have done that, but this is just the test app file, not used by production code

File: src/LinkGroup.js (hunk 3 above)
Comment (Tri4Ty) [reply src/LinkGroup.js]: proper coding standards... you should read up on the difference between let and const https://tylermcginnis.com/var-let-const/

File: src/LinkGroup.js (hunk 5 above)
Comment (Tri4Ty) [reply src/LinkGroup.js]: it was a custom prototype validation... since we moved this to accept nulls, the custom type wasn't needed anymore... also, it was causing lint errors and since it's not needed, I cleaned up the code to ensure clean builds (Boy Scout rule)

File: src/LinkGroup.js (hunk 2 above)
Comment (Tri4Ty) [reply src/LinkGroup.js]: noting is selected by default ... which is what undefined signifies

File: src/Link.js (hunk 1 above)
Comment (Tri4Ty) [reply src/Link.js]: same answer as LinkGroup

File: LICENSE [hunk 19]
```diff
@@ -6,60 +6,60 @@
 
    1. Definitions.
 
-      "License" shall mean the terms and conditions for use, reproduction,
```
[the remark this answers is not in the document]
Comment (Tri4Ty) [reply LICENSE]: ya it was weird ... I guess we didn't have License file in the ignore file. I'm not crazy about it, but it doesn't impact anything and it was done through automation... no harm, no foul

File: src/LinkGroup.js (hunk 4 above)
Comment (Tri4Ty) [reply src/LinkGroup.js]: you should try reading a little further down... seems there's a return statement right there

File: package.json (hunk 15 above)
Comment (FrancisPaquette) [inline package.json]: Have you spent time thinking about putting this in peer-dependencies?
Otherwise, each module will be bringing this along (and possibly with multiple versions over time).

File: src/Link.js [hunk 20]
```diff
@@ -21,18 +31,13 @@ const Link = props => {
...
-    if (propValue === null) return;
-    if (typeof propValue === "string") return;
-    return new Error(`${componentName} only accepts null or string`);
-  },
+  id: PropTypes.string,
   displayName: PropTypes.string.isRequired,
   value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
   selected: PropTypes.bool,
-  selectionCallback: PropTypes.func.isRequired
+  selectionCallback: PropTypes.func.isRequired,
```
Comment (FrancisPaquette) [inline src/Link.js]: The comma at the end seems strange.
Can you explain to me?  (Since the next line has "};")

File: .storybook/webpack.config.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (FrancisPaquette) [reply .storybook/webpack.config.js]: Correct me if I am wrong Steve but all the changes from double quote to single quote was done automatically by prettier, right?
ie It is not something that a developer has to remind himself constantly when coding?

File: dev/src/App.js (hunk 8 above)
Comment (sali1982) [reply dev/src/App.js]: It is okay, but even in test and dev files it is good to try using clear names.

File: package.json (hunk 15 above)
[the remark this answers is not in the document]
Comment (sali1982) [reply package.json]: I am not asking about the dependency to go read about it :)
I am asking what does this specific dependency is doing  :)

File: src/Link.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (sali1982) [reply src/Link.js]: it is good before actually doing dev for this, I got to learn it ;)

File: src/Link.js (hunk 20 above)
Comment (sali1982) [reply src/Link.js]: looks like the new eslint forses? Please explain if I am wrong.

File: src/LinkGroup.js (hunk 2 above)
Comment (sali1982) [reply src/LinkGroup.js]: I am suggesting having first one selected unless deifned in props the others. maybe

File: src/LinkGroup.js (hunk 4 above)
Comment (sali1982) [reply src/LinkGroup.js]: I do not see any return for the linkList itself. It used to be:
let linkList = links.map(link => {
      return (
        <Link
          key={link.id}
          {...link}
          selected={link.id === selected}
          selectionCallback={this.linkSelection}
        />
      );
return <ul className="react-link-group">{linkList}</ul>;

and changed to:
const linkList = links.map((link) => (
   <Link
        key={link.id}
        {...this.props}
        {...link}
        selected={link.id === selected}
        selectionCallback={this.linkSelection}
      />
    ));
return <ul className={theme.reactLinkGroup}>{linkList}</ul>;
Whic is interesting to know {linkList} works withouth the return.

File: package.json (hunk 15 above)
Comment (Tri4Ty) [reply package.json]: I did think about this and I struggled with it. On one hand, for our environment/platform, all components will be using css-modules and thus having a peer-dependency makes sense (Mosaic will bring it in). On the other hand, this is a generic/reusable/shared component which anyone can pick up... do we want to enforce them to have to use theming? I went with not forcing people to have to use theming so the ability to reuse this component is easier. If we see there are some bundle sizing issues we need to reduce, we can revisit this.

File: src/Link.js (hunk 20 above)
Comment (Tri4Ty) [reply src/Link.js]: yes, Salma is correct ... it's "proper" JavaScript according to standards. Do I agree with it fully... not really, but I was able to configure Prettier to do this for us, so not an issue as it will get applied for us
